streams/src/main/java/org/apache/kafka/streams/kstream/internals/AbstractKStreamTimeWindowAggregateProcessor.java

```diff
@@ -0,0 +1,170 @@
+                                                          final EmitStrategy emitStrategy,
+                                                          final boolean sendOldValues) {
+        this.storeName = storeName;
+        this.emitStrategy = emitStrategy;
+        this.sendOldValues = sendOldValues;
+    }
+
+    @Override
+    public void init(final ProcessorContext<Windowed<KIn>, Change<VAgg>> context) {
+        super.init(context);
+        internalProcessorContext = (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
+        final StreamsMetricsImpl metrics = internalProcessorContext.metrics();
+        final String threadId = Thread.currentThread().getName();
+        droppedRecordsSensor = droppedRecordsSensor(threadId, context.taskId().toString(), metrics);
+        emittedRecordsSensor = emittedRecordsSensor(threadId, context.taskId().toString(),
+            internalProcessorContext.currentNode().name(), metrics);
+        emitFinalLatencySensor = emitFinalLatencySensor(threadId, context.taskId().toString(),
+            internalProcessorContext.currentNode().name(), metrics);
+        windowStore = context.getStateStore(storeName);
+
+        if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+            // Restore last emit close time for ON_WINDOW_CLOSE strategy
+            final Long lastEmitTime = internalProcessorContext.processorMetadataForKey(storeName);
+            if (lastEmitTime != null) {
+                lastEmitWindowCloseTime = lastEmitTime;
+            }
+            final long emitInterval = StreamsConfig.InternalConfig.getLong(
+                context.appConfigs(),
+                EMIT_INTERVAL_MS_KSTREAMS_WINDOWED_AGGREGATION,
+                1000L
+            );
+            timeTracker.setEmitInterval(emitInterval);
+        } else {
+            tupleForwarder = new TimestampedTupleForwarder<>(
+                windowStore,
+                context,
+                new TimestampedCacheFlushListener<>(context),
+                sendOldValues);
+        }
+    }
+
+    protected void maybeForwardUpdate(final Record<KIn, VIn> record,
+                                      final Window window,
+                                      final VAgg oldAgg,
+                                      final VAgg newAgg,
+                                      final long newTimestamp) {
+        if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+            return;
+        } else if (tupleForwarder == null) {
+            throw new IllegalStateException("Emit strategy type is " + emitStrategy.type() + " but flush listener is not initialized.");
+        }
+
+        tupleForwarder.maybeForward(
+            record.withKey(new Windowed<>(record.key(), window))
+                .withValue(new Change<>(newAgg, sendOldValues ? oldAgg : null))
+                .withTimestamp(newTimestamp));
+    }
+
+    protected boolean shouldEmitFinal(final long closeTime) {
```

Comment:
It seems this method is called as a first step inside the implementation of `maybeForwardFinalResult` -- thus, I am wondering if we should not make it private and call inside `maybeMeasureEmitFinalLatency` ?